content/en/blog/_posts/2024-02-29-cri-o-seccomp-oci-artifacts.md

```diff
@@ -0,0 +1,486 @@
+Events:
+  Type     Reason     Age                 From               Message
+  ----     ------     ----                ----               -------
+  Normal   Scheduled  117s                default-scheduler  Successfully assigned default/pod to 127.0.0.1
+  Normal   Pulling    117s                kubelet            Pulling image "nginx:1.25.3"
+  Normal   Pulled     111s                kubelet            Successfully pulled image "nginx:1.25.3" in 5.948s (5.948s including waiting)
+  Warning  Failed     7s (x10 over 111s)  kubelet            Error: setup seccomp: unable to load local profile "/var/lib/kubelet/seccomp/nginx-1.25.3.json": open /var/lib/kubelet/seccomp/nginx-1.25.3.json: no such file or directory
+  Normal   Pulled     7s (x9 over 111s)   kubelet            Container image "nginx:1.25.3" already present on machine
+```
+
+The major obstacle of having to manually distribute the `Localhost` profiles
+will lead many end-users to fall back to `RuntimeDefault` or even running their
+workloads as `Unconfined` (with disabled seccomp).
+
+## CRI-O to the rescue
+
+The Kubernetes container runtime [CRI-O](https://github.com/cri-o/cri-o)
+provides various features using custom annotations. The v1.30 release
+[adds](https://github.com/cri-o/cri-o/pull/7719) support for a new set of annotations
+called `seccomp-profile.kubernetes.cri-o.io/POD` and
+`seccomp-profile.kubernetes.cri-o.io/<CONTAINER>`. Those annotations allow you to
+specify:
+
+- a seccomp profile for a specific container, when used as:
+  `seccomp-profile.kubernetes.cri-o.io/<CONTAINER>` (example: `seccomp-profile.kubernetes.cri-o.io/webserver: 'registry.example/example/webserver:v1'`)
+- a seccomp profile for every container within a pod, when used without the
+  container name suffix but the reserved name `POD`:
+  `seccomp-profile.kubernetes.cri-o.io/POD`
+- a seccomp profile for a whole container image, if the image itself contains
+  the annotation `seccomp-profile.kubernetes.cri-o.io/POD` or
+  `seccomp-profile.kubernetes.cri-o.io/<CONTAINER>`.
+
+CRI-O will only respect the annotation if the runtime is configured to allow it,
+as well as for workloads running as `Unconfined`. All other workloads will still
+use the value from the `securityContext` with a higher priority.
+
+The annotations alone will not help much with the distribution of the profiles,
+but the way they can be referenced will! For example, you can now specify
+seccomp profiles like regular container images by using OCI artifacts:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: pod
+  annotations:
+    seccomp-profile.kubernetes.cri-o.io/POD: quay.io/crio/seccomp:v1
+spec: …
+```
+
+The image `quay.io/crio/seccomp:v1` contains a `seccomp.tar` tarball, which
+contains the actual profile. Tools like [ORAS](https://oras.land) or
+[Skopeo](https://github.com/containers/skopeo) can be used to inspect the
+contents of the artifact:
```

Comment:
Is this:
- an OCI image and therefore also an OCI object
- an OCI object, but not actually a container image
- a prototype for a future approach that will be different from the prototype?

I'd like to make that clear to readers, too.

Reply:
It's basically just an image, I'm going to refer to that.